4 - AnimalShelter/Code/AnimalShelter.cs

```diff
@@ -0,0 +1,81 @@
+﻿using AnimalShelterCode;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Linq;
+
+namespace AnimalShelterCode
+{
+    public class AnimalShelter
+    {
+        public int AnimalCounter { get; private set; }
+
+        public List<Animal> Animals { get; private set; }
+
+        public AnimalShelter()
+        {
+            Animals = new List<Animal>();
+            AnimalCounter = 1;
+        }
+
+        public OperationResult AddAnimal(Animal animal)
+        {
+            if (animal.Type == AnimalType.Bear || animal.Type == AnimalType.Deer) 
+            { 
+                return new OperationResult($"Animal Type {animal.Type} is not Supported");
+            }
+            else
+            {
+                animal.UniqueAnimalId = AnimalCounter++;
+                Animals.Add(animal);
+                return new OperationResult(animal);
+            }
+        }
+
+        public List<Animal> GetAnimals(AnimalType filtertype)
+        {
+            var filteredlist = from animal in Animals
+                               where animal.Type == filtertype
+                               select animal;
+
+            return filteredlist.ToList();
+        }
+        public List<Animal> GetAnimals()
+        {
+            return Animals;
+        }
+
+        public OperationResult RemoveAnimal(Animal animal)
+        {
+            if (Animals.Contains(animal))
+            {
+                Animals.Remove(animal);
+                return new OperationResult(animal);
+            }
+            else
+            {
+                return new OperationResult("Animal is not present in shelter");
+            }
+        }
```

Comment:
The requirements were rather complex for this one but I am glad you saw through them to realize that if you store your animal state properly, this is all you have to do :) - good job!